server/src/main/java/org/spine3/server/storage/RecordStorage.java

```diff
@@ -115,22 +118,36 @@ public void write(Map<I, EntityRecord> records) {
         writeRecords(records);
     }
 
-    /**
-     * Marks the record with the passed ID as {@code archived}.
-     *
-     * @param id the ID of the record to mark
-     */
-    public abstract void markArchived(I id);
+    @Override
+    public Optional<Visibility> readVisibility(I id) {
+        final Optional<EntityRecord> optional = read(id);
+        if (optional.isPresent()) {
+            return Optional.of(optional.get().getVisibility());
+        }
+        return Optional.absent();
+    }
 
-    /**
-     * Marks the record with the passed ID as {@code deleted}.
-     *
-     * <p>This method does not delete the record.
-     * To delete the record please call {@link #delete(Object)}
-     *
-     * @param id the ID of the record to mark
-     */
-    public abstract void markDeleted(I id);
+    @Override
+    public void writeVisibility(I id, Visibility visibility) {
+        final Optional<EntityRecord> optional = read(id);
+        if (optional.isPresent()) {
+            final EntityRecord record = optional.get();
+            final EntityRecord updated = record.toBuilder()
+                                               .setVisibility(visibility)
+                                               .build();
+            write(id, updated);
+        } else {
+            // The AggregateStateId is a special case, which is not handled by the Identifier class.
```

Comment:
The code that creates the string form of an ID is actually in the `Identifier.Type` enum. 

The `AggregateStateId` being a plain Java class (not one of primitive wrappers that we support, and not a `Message`) doesn't fall into the `case` in `Identifier.toString()`, which causes the error.

I didn't want to extend the `Identifier` for this particular case, hence the comment.